src/tracker.rs

```diff
@@ -0,0 +1,300 @@
+    pub fn new() -> Self {
+        Self {
+            count_threshold: DEFAULT_TAG_THRESHOLD,
+            actions: Vec::new(),
+        }
+    }
+
+    /// Configure the tracker to emit an event when the threshold is reached.
+    /// The count of unique tag sets, per metric, is provided as the tags
+    /// for the event, i.e. for a given metric `metric`, there will be a tag `metric:count`,
+    /// where count is the number of unique tag sets seen for that metric.
+    /// Any number of events may be configured, and all will be emitted when the
+    /// threshold is reached.
+    pub fn with_event(mut self, title: String, text: String) -> Self {
+        self.actions.push(ThresholdAction::Event { title, text });
+        self
+    }
+
+    /// Add a custom action to execute when the custom metric threshold is reached.
+    /// These actions are run exactly once, at the point the threshold is reached.
+    /// Any number of actions may be added.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// prima_datadog::TagTrackerConfiguration::new().with_custom_action(|metric, tags| {
+    ///     println!("Exceeded custom metric threshold for metric {} with tags {:?}", metric, tags);
+    /// });
+    /// ```
+    pub fn with_custom_action(mut self, custom_action: impl FnMut(&str, &[&str]) + Send + Sync + 'static) -> Self {
+        self.actions
+            .push(ThresholdAction::Custom(Box::new(custom_action) as Box<_>));
+        self
+    }
+
+    /// Set the threshold for the maximum number of custom metrics
+    /// This defaults to ```DEFAULT_TAG_THRESHOLD```
+    ///
+    /// The threshold is the maximum number of "custom metrics" which can
+    /// be generated before the tracker actions are run.
+    ///
+    /// A "custom metric" is defined as the unique combination of metric name
+    /// and unique tag set, e.g., a metric, "test", with the following tag sets:
+    /// - [a, b, c]
+    /// - [a, b, d]
+    /// - [a, c, d]
+    /// is counted as 3 "custom metrics", as there are 3 unique tag sets here.
+    ///
+    /// See https://docs.datadoghq.com/developers/metrics/custom_metrics/ for
+    /// more information.
+    pub fn with_threshold(mut self, count_threshold: usize) -> Self {
+        self.count_threshold = count_threshold;
+        self
+    }
+
+    pub(crate) fn build(self) -> Tracker {
+        Tracker::new(self.count_threshold, self.actions)
+    }
+}
+
+/// An iterator over tags, which can be rewound to the beginning.
+struct RewindableTagsIter<'a> {
+    tags: &'a mut dyn Iterator<Item = &'a str>,
+    cursor: usize,
+    len: usize,
+}
+impl RewindableTagsIter<'_> {
+    fn rewind(&mut self) -> &mut Self {
+        for _ in 0..(self.len - self.cursor) {
+            let _next = self.tags.next();
+            debug_assert!(_next.is_some(), "The `tags` in `RewindableTagsIter` should never stop returning items, use `.cycle()` to make it infinite");
+        }
+        self.cursor = 0;
+        self
+    }
+}
```

Comment:
I think we could do this more cleanly by copying what the Cycle iter does in std(https://doc.rust-lang.org/src/core/iter/adapters/cycle.rs.html#13), and adding a "rewind" method.

Reply:
That would make way more sense

Reply:
See above, removed for now